src/Compilers/CSharp/Portable/Binder/PatternSwitchBinder.cs

```diff
@@ -12,51 +14,131 @@ namespace Microsoft.CodeAnalysis.CSharp
-            var locals = GetDeclaredLocalsForScope(node);
-            var functions = GetDeclaredLocalFunctionsForScope(node);
+            ImmutableArray<BoundPatternSwitchSection> switchSections = BindPatternSwitchSections(originalBinder, diagnostics, out BoundPatternSwitchLabel defaultLabel);
+            ImmutableArray<LocalSymbol> locals = GetDeclaredLocalsForScope(node);
+            ImmutableArray<LocalFunctionSymbol> functions = GetDeclaredLocalFunctionsForScope(node);
+            BoundDecisionDag decisionDag = DecisionDagBuilder.CreateDecisionDag(
+                compilation: this.Compilation,
+                syntax: node,
+                switchGoverningExpression: boundSwitchGoverningExpression,
+                switchSections: switchSections,
+                // If there is no explicit default label, the default action is to break out of the switch
+                defaultLabel: defaultLabel?.Label ?? BreakLabel,
+                diagnostics);
+
+            // Report subsumption errors, but ignore the input's constant value for that.
+            CheckSwitchErrors(node, boundSwitchGoverningExpression, ref switchSections, decisionDag, diagnostics);
+
+            if (boundSwitchGoverningExpression.ConstantValue != null)
+            {
+                // When the input is constant, we use that to reshape the decision dag that is returned
+                // so that flow analysis will see that some of the cases may be unreachable.
+                decisionDag = decisionDag.SimplifyDecisionDagForConstantInput(boundSwitchGoverningExpression, Conversions, diagnostics);
+            }
+
             return new BoundPatternSwitchStatement(
-                node, boundSwitchGoverningExpression, someCaseMatches,
-                locals, functions, switchSections, defaultLabel, this.BreakLabel, isComplete);
+                syntax: node,
+                expression: boundSwitchGoverningExpression,
+                innerLocals: locals,
+                innerLocalFunctions: functions,
+                switchSections: switchSections,
+                defaultLabel: defaultLabel,
+                breakLabel: this.BreakLabel,
+                decisionDag: decisionDag);
+        }
+
+        private void CheckSwitchErrors(
+            SwitchStatementSyntax node,
+            BoundExpression boundSwitchGoverningExpression,
+            ref ImmutableArray<BoundPatternSwitchSection> switchSections,
+            BoundDecisionDag decisionDag,
+            DiagnosticBag diagnostics)
+        {
+            HashSet<LabelSymbol> reachableLabels = decisionDag.ReachableLabels;
+            bool isSubsumed(BoundPatternSwitchLabel switchLabel)
+            {
+                return !reachableLabels.Contains(switchLabel.Label);
+            }
+
+            // If no switch sections are subsumed, just return
+            if (switchSections.Count(s => s.SwitchLabels.Count(l => isSubsumed(l)) != 0) == 0)
```

Comment:
>switchSections.Count [](start = 16, length = 20)

Consider using `Any` or `All` to avoid walking all sections always: `!switchSections.Any(s => s.SwitchLabels.Any(l => isSubsumed(l)))`

Reply:
In the normal (non-error case) we'll be visiting every single label. We'd only be improving things for the (rare) error case. But I agree the change is worth doing.

Reply:
There is no `Any` method for immutable arrays.

Reply:
`System.Linq.ImmutableArrayExtensions.Any<T>(this ImmutableArray<T>, Func<T, bool>)`

Reply:
Addressed in [https://github.com/dotnet/roslyn/pull/25662](https://github.com/dotnet/roslyn/pull/25662) or earler.

---
In reply to: [175854338](https://github.com/dotnet/roslyn/pull/25544#discussion_r175854338) [](ancestors = 175854338)